### PR TITLE
Offset Bug Fix for OP Indexer

### PR DESCRIPTION
**Pull request by ethenotethan**
**Description**
Fixed two bugs:
1. Failed `DepositTx` where mint = nil 
2. Failed L1 starting height bug where bridge processor would try starting at non-indexed height and continuously fail 

Updated documentation 
**Tests**
N/A

**Additional context**

Add any other context about the problem you're solving.

**Metadata**

- Fixes #[Link to Issue]

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (hamdiallam) [inline .vscode/launch.json]: rm these changes

File: indexer/migrations/20230523_create_schema.sql [hunk 2]
```diff
@@ -16,6 +16,13 @@ END $$;
  * BLOCK DATA
  */
 
+
```
Comment (hamdiallam) [inline indexer/migrations/20230523_create_schema.sql]: I think these were unintended

File: indexer/database/blocks.go [hunk 3]
```diff
@@ -174,12 +175,14 @@ func (db *blocksDB) LatestObservedEpoch(fromL1Height *big.Int, maxL1Range uint64
...
 	}
 
-	// Lower Bound (the default `fromTimestamp = 0` suffices genesis representation)
+	// Lower Bound (the default `fromTimestamp = l1_starting_heigh` (default=0) suffices genesis representation)
 	if fromL1Height.BitLen() > 0 {
 		var header L1BlockHeader
 		result := db.gorm.Where("number = ?", fromL1Height).Take(&header)
 		if result.Error != nil {
 			if errors.Is(result.Error, gorm.ErrRecordNotFound) {
+				log.Warn("Could not fetch latest L1 block header in bridge processor", "number", fromL1Height,
```
Comment (hamdiallam) [inline indexer/database/blocks.go]: rm these log lines. You can push these into the bridge processer though when it calls `LatestObservedEpoch`

File: indexer/database/blocks.go [hunk 4]
```diff
@@ -196,21 +199,26 @@ func (db *blocksDB) LatestObservedEpoch(fromL1Height *big.Int, maxL1Range uint64
 			l1QueryFilter = fmt.Sprintf("%s AND number <= %d", l1QueryFilter, maxHeight)
 		}
 
+		// Fetch most recent header from l1_block_headers table
 		var l1Header L1BlockHeader
 		result := db.gorm.Where(l1QueryFilter).Order("timestamp DESC").Take(&l1Header)
 		if result.Error != nil {
 			if errors.Is(result.Error, gorm.ErrRecordNotFound) {
+				log.Warn("Could not fetch latest L1 block header in bridge processor", "from_timestamp",
```
Comment (hamdiallam) [inline indexer/database/blocks.go]: same here

**Review by hamdiallam — COMMENTED** [review]
thanks for the fix! some comments on the extra changes

File: indexer/database/blocks.go (hunk 4 above)
Comment (ethenotethan) [reply indexer/database/blocks.go]: [a86737d](https://github.com/ethereum-optimism/optimism/pull/7649/commits/a86737dfd3ef8961c485586d7b9f2b790a67f460)

File: indexer/database/blocks.go (hunk 3 above)
Comment (ethenotethan) [reply indexer/database/blocks.go]: [a86737d](https://github.com/ethereum-optimism/optimism/pull/7649/commits/a86737dfd3ef8961c485586d7b9f2b790a67f460)

File: indexer/migrations/20230523_create_schema.sql (hunk 2 above)
Comment (ethenotethan) [reply indexer/migrations/20230523_create_schema.sql]: [a86737d](https://github.com/ethereum-optimism/optimism/pull/7649/commits/a86737dfd3ef8961c485586d7b9f2b790a67f460)

File: .vscode/launch.json (hunk 1 above)
Comment (ethenotethan) [reply .vscode/launch.json]: [a86737d](https://github.com/ethereum-optimism/optimism/pull/7649/commits/a86737dfd3ef8961c485586d7b9f2b790a67f460)

File: indexer/database/blocks.go [hunk 5]
```diff
@@ -170,14 +169,11 @@ func (db *blocksDB) LatestObservedEpoch(fromL1Height *big.Int, maxL1Range uint64
...
 	var fromTimestamp, toTimestamp uint64
 
-	if fromL1Height == nil {
-		fromL1Height = bigint.Zero
-	}
-
-	// Lower Bound (the default `fromTimestamp = 0` suffices genesis representation)
-	if fromL1Height.BitLen() > 0 {
-		var header L1BlockHeader
+	// Lower Bound (the default `fromTimestamp = l1_starting_heigh` (default=0) suffices genesis representation)
```
Comment (hamdiallam) [inline indexer/database/blocks.go]:
```suggestion
	// Lower Bound (the default `fromTimestamp = l1_starting_height` (default=0) suffices genesis representation)
```

File: indexer/etl/l1_etl.go [hunk 6]
```diff
@@ -131,7 +131,14 @@ func (l1Etl *L1ETL) Start(ctx context.Context) error {
 						return err
 					}
 					return nil
-				}); err != nil {
+				})
+
+				if err != nil && strings.Contains(err.Error(), "duplicate key value violates unique constraint") {
```
Comment (hamdiallam) [inline indexer/etl/l1_etl.go]: I don't think we should silently ignore this as it's indicative of other errors, rather always ensure that the indexer is never re-indexing the same data.

File: indexer/migrations/20230523_create_schema.sql (hunk 2 above)
Comment (hamdiallam) [reply indexer/migrations/20230523_create_schema.sql]: Looks like there were re-added

File: indexer/processors/contracts/optimism_portal.go [hunk 7]
```diff
@@ -64,14 +64,19 @@ func OptimismPortalTransactionDepositEvents(contractAddress common.Address, db *
 			return nil, err
 		}
 
+		mint := depositTx.Mint
+		if mint == nil {
+			mint = big.NewInt(0)
```
Comment (hamdiallam) [inline indexer/processors/contracts/optimism_portal.go]:
```suggestion
			mint = bigint.Zero
```

File: indexer/node/header_traversal.go [hunk 8]
```diff
@@ -15,6 +15,7 @@ var (
 )
 
 type HeaderTraversal struct {
+	started   bool
```
Comment (hamdiallam) [inline indexer/node/header_traversal.go]: unused

Comment (ethenotethan) [reply indexer/node/header_traversal.go]: d605488

File: indexer/processors/contracts/optimism_portal.go (hunk 7 above)
Comment (ethenotethan) [reply indexer/processors/contracts/optimism_portal.go]: d605488

File: indexer/etl/l1_etl.go (hunk 6 above)
Comment (ethenotethan) [reply indexer/etl/l1_etl.go]: d605488

**Review by hamdiallam — APPROVED** [review]

**Pull request merged**